source/controllers/PPSaveController.cpp

```diff
@@ -0,0 +1,109 @@
+#include "PPSaveController.h"
+
+#define SAVE_PATH (Application::get()->getSaveDirectory() + "save")
+
+SaveController::LevelMetadata SaveController::_getLevel(
+    const string &level) const {
+    auto i = _levels.find(level);
+    if (i == _levels.end()) return LevelMetadata();
+    return i->second;
+}
+
+SaveController::LevelMetadata &SaveController::_ensureLevel(
+    const string &level) {
+    return _levels[level];
+}
+
+void SaveController::_load() {
+    auto r = JsonReader::alloc(SAVE_PATH);
+    if (r == nullptr) return;
+    json_t v = r->readJson();
+    if (v->getInt("version", 1) == 1) {
+        _colorblind = v->getBool("colorblind", false);
+        _sfxVolume = v->getFloat("sfxVolume", 1);
+        _bgmVolume = v->getFloat("bgmVolume", 1);
```

Comment:
can be done in the future but maybe there can be a setting to set whether the palette is on the right or on the left?